src/nomad_simulations/schema_packages/data_types.py

```diff
@@ -0,0 +1,260 @@
+    Bounds are by default +/- infinity (`float('inf')`, `float('-inf')` respectively).
+    """
+
+    _error_message_prefix: str = 'All values'
+
+    def __init__(self) -> None:
+        self._min_value = float('-inf')
+        self._max_value = float('inf')
+        self._min_inclusive = True
+        self._max_inclusive = True
+
+    def min(self: T, value: Union[int, float], *, inclusive: bool = True) -> T:
+        """Set minimum bound. Supports +/- infinity."""
+        self._min_value = value
+        self._min_inclusive = inclusive
+        return self
+
+    def max(self: T, value: Union[int, float], *, inclusive: bool = True) -> T:
+        """Set maximum bound. Supports +/- infinity."""
+        self._max_value = value
+        self._max_inclusive = inclusive
+        return self
+
+    @abstractmethod
+    def _filter_values_for_validation(self, flat_values: list[Any]) -> list[Any]:
+        """Filter values for validation (e.g., remove NaN for floats)."""
+        pass
+
+    def _validate_bounds(self, normalized_value: Any) -> None:
+        """Common bounds validation logic."""
+        flat_values = list(_flatten_values(normalized_value))
+        valid_values = self._filter_values_for_validation(flat_values)
+
+        if valid_values:
+            invalid_values = [
+                v
+                for v in valid_values
+                if not _check_bounds(
+                    v,
+                    self._min_value,
+                    self._max_value,
+                    self._min_inclusive,
+                    self._max_inclusive,
+                )
+            ]
+
+            if invalid_values:
+                min_val = min(valid_values)
+                max_val = max(valid_values)
+                bounds = _get_bounds_str(
+                    self._min_value,
+                    self._max_value,
+                    self._min_inclusive,
+                    self._max_inclusive,
+                )
+                raise ValueError(
+                    f'{self._error_message_prefix} must be in {bounds}, got range [{min_val}, {max_val}]'
+                )
+
+
+class BoundedInt(BoundedDataType, ExactNumber):
+    """
+    32-bit integer data type with configurable bounds (bound value and inclusion).
+    Bounds are by default +/- infinity (`float('inf')`, `float('-inf')` respectively).
+
+    Precision: 32-bit signed integer (range: -2,147,483,648 to 2,147,483,647)
+    Convertible from: Python int, np.int32, np.int16, np.int8
+
+    Examples:
+        BoundedInt().min(1)                    # >= 1 (positive integers)
+        BoundedInt().min(0)                    # >= 0 (non-negative integers)
+        BoundedInt().min(1, inclusive=False)   # > 1
+        BoundedInt().max(10)                   # <= 10
+        BoundedInt().min(1).max(10)            # 1 <= value <= 10
+    """
+
+    __slots__ = ('_min_value', '_max_value', '_min_inclusive', '_max_inclusive')
```

Comment:
This line shall go to the parent?

Reply:
If I rework the interface, I only need 1 class, so this is resolved then.